### PR TITLE
Example using a PIO interrupt

### LINKED ISSUE #377: Add an example using PIO interrupts
Not much makes use of interrupts from PIO. An example using the uart pio would be useful

### PR DESCRIPTION
**Pull request by peterharperuk**
Add PIO UART receive example which uses interrupts.

Fixes #377

### CODE REVIEW

**Comment by peterharperuk** [comment]
How's that!

**Comment by andygpz11** [comment]
I guess this could be done at merge (In which case I'll mark it as
reviewed) or you could make one tiny change and push or even use commit
--amend, edit the comment and then push -f

**Comment by peterharperuk** [comment]
It should be fixed now.

**Review by andygpz11 — APPROVED** [review]

File: pio/uart_rx/uart_rx_intr.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#include "pico/async_context_threadsafe_background.h"
+
+#include "hardware/pio.h"
+#include "hardware/uart.h"
+#include "uart_rx.pio.h"
+
+// This program
+// - Uses UART1 (the spare UART, by default) to transmit some text
+// - Uses a PIO state machine to receive that text
+// - Use an interrupt to determine when the PIO FIFI has some data
```
Comment (lurch) [inline pio/uart_rx/uart_rx_intr.c:22]: FIFI -> FIFO ?

File: pio/uart_rx/uart_rx_intr.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        if (irq_get_exclusive_handler(pio_irq)) {
+            panic("All IRQs are in use");
+        }
+    }
+
+    // Enable interrupt
+    irq_add_shared_handler(pio_irq, pio_irq_func, PICO_SHARED_IRQ_HANDLER_DEFAULT_ORDER_PRIORITY); // Add a shared IRQ handler
+    irq_set_enabled(pio_irq, true); // Enable the IRQ
+    const uint irq_index = pio_irq - ((pio == pio0) ? PIO0_IRQ_0 : PIO1_IRQ_0); // Get index of the IRQ
+    pio_set_irqn_source_enabled(pio, irq_index, pis_sm0_rx_fifo_not_empty + sm, true); // Set pio to tell us when the fifi is NOT empty
```
Comment (lurch) [inline pio/uart_rx/uart_rx_intr.c:147]: Another "fifi" here

**Review by andygpz11 — APPROVED** [review]

**Review by kilograham — APPROVED** [review]

**Pull request merged**